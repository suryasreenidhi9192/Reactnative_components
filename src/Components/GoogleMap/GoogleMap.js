/**
*
* GoogleMap
*
*/
// @flow
import React, { Component } from 'react';
import ReactDOMServer from 'react-dom/server';
import findIndex from 'lodash/findIndex';
import styles from './GoogleMap.style';
import GoogleMapInfoWindow from '../GoogleMapInfoWindow';
import withStyles from '../../../lib/withStyles';
import { STATIC_IMAGE } from '../../../constants';

export class GoogleMap extends Component<Props> {
  state = {
    activeInfoWindow: null,
  };

  componentDidMount() {
    this.initMap();
  }

  onMarkerClick = ({
    isStoreLocator,
    store,
    marker,
    infoWindow,
    map,
  }) => {
    const {
      activeInfoWindow,
    } = this.state;
    if (!isStoreLocator) {
      const storesListTop = document.querySelector('.stores-list').offsetTop;
      const storeToScroll = document.getElementById(`store-${store.storeId}`).offsetTop;
      const toScroll = storeToScroll - storesListTop;
      document.getElementsByClassName('stores-list')[0].scrollTop = toScroll;
    } else {
      if (activeInfoWindow) {
        activeInfoWindow.close();
      }
      this.setState({
        activeInfoWindow: infoWindow,
      });
      infoWindow.open(map, marker);
    }
  };

  onMapClicked = () => {
    const {
      activeInfoWindow,
    } = this.state;
    if (this.state.activeInfoWindow) {
      activeInfoWindow.close();
      this.setState({
        activeInfoWindow: null,
      });
    }
  };

  initMap = () => {
    if (!this.mapRef) return;
    const { google } = window;
    const {
      storeDetails,
      lat,
      lng,
      isStoreLocator,
      stores,
      deviceType,
      activeBrand,
    } = this.props;
    const { allStores } = storeDetails;
    const options = {
      center: lat && lng ? new google.maps.LatLng(lat, lng) : null,
      zoom: 10,
      zoomControl: true,
      mapTypeControl: true,
      mapTypeControlOptions: null, // default defined after including API
      streetViewControl: true,
      streetViewControlOptions: null,
      scaleControl: true,
      panControl: true,
      mapTypeId: google.maps.MapTypeId.ROADMAP,
    };
    const map = new google.maps.Map(this.mapRef, options);

    map.addListener('click', () => {
      this.onMapClicked();
    });

    const storeList = isStoreLocator ? allStores : stores;
    let counter = 0;
    storeList.forEach((store) => {
      const storeId = !isStoreLocator ? store.storeId : store.id;
      const { yextRoutableLat, yextRoutableLng } = store;
      const storeIndex = findIndex(allStores, o => o && o.id === storeId);
      if (storeIndex < 0) {
        return;
      }
      const storeDetail = isStoreLocator
        ? store
        : allStores[storeIndex];
      counter += 1;
      const storeLat = parseFloat(yextRoutableLat);
      const storeLng = parseFloat(yextRoutableLng);
      const marker = new google.maps.Marker({
        position: { lat: storeLat, lng: storeLng },
        map,
        icon: `${STATIC_IMAGE}/map_markers/${activeBrand}/marker_${counter}.png`,
      });
      const infoWindow = new google.maps.InfoWindow({
        content: ReactDOMServer.renderToString(<GoogleMapInfoWindow
          store={storeDetail}
          storeProperties={storeDetail}
          deviceType={deviceType}
          activeBrand={activeBrand}
          isStoreLocator={isStoreLocator}
        />),
      });
      marker.addListener('click', () => {
        this.onMarkerClick({
          isStoreLocator,
          store,
          marker,
          infoWindow,
          map,
        });
      });
    });
  };

  render() {
    const {
      storeDetails,
    } = this.props;
    const { allStores } = storeDetails;
    if (!allStores) {
      return null;
    }

    const style = {
      minWidth: '100%',
      minHeight: '100%',
    };

    return (
      <div
        ref={(el) => {
          if (!this.mapRef) {
            this.mapRef = el;
            this.initMap();
          }
        }}
        style={style}
      />
    );
  }
}

export default withStyles(GoogleMap, styles);
export { GoogleMap as GoogleMapVanilla };
