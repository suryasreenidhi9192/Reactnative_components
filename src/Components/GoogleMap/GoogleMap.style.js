import { css } from 'styled-components';

import { DESKTOP, MOBILE } from '../../../constants';

const commonStyles = css`
  .map-list {
    .map-marker {
      color: white;
    }
  }
`;

export default {
  [DESKTOP]: css`
    ${commonStyles};
  `,
  [MOBILE]: css`
    ${commonStyles};
  `,
};
