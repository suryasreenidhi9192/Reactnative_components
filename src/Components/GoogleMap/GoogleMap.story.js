import React from 'react';
import { storiesOf } from '@storybook/react';
// import { action } from '@storybook/addon-actions';
import GoogleMap from './index';

storiesOf('GoogleMap', module)
  .add('', () => (<GoogleMap>Hello GoogleMap</GoogleMap>));
