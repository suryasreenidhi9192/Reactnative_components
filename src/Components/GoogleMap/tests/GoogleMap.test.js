import { fromJS } from 'immutable';
import { shallow, mount } from 'enzyme';
// import renderer from 'react-test-renderer';
import 'jest-styled-components';

import { GoogleMapVanilla } from '../GoogleMap';
import { DESKTOP, OUTLET_LOFT } from '../../../../constants';

/* eslint-disable */
const location = fromJS({
  results: [
    {
      address_components: [
        {
          long_name: '10020',
          short_name: '10020',
          types: [
            'postal_code',
          ],
        },
      ],
      formatted_address: 'New York, NY 10020, USA',
      geometry: {
        bounds: {
          northeast: {
            lat: 40.76171890000001,
            lng: -73.97720079999999,
          },
          southwest: {
            lat: 40.757284,
            lng: -73.984174,
          },
        },
        location: {
          lat: 40.7605505,
          lng: -73.98226830000002,
        },
        location_type: 'APPROXIMATE',
        viewport: {
          northeast: {
            lat: 40.76171890000001,
            lng: -73.97720079999999,
          },
          southwest: {
            lat: 40.757284,
            lng: -73.984174,
          },
        },
      },
      place_id: 'ChIJgbqqxP5YwokRfAdWTsd-vEg',
      types: [
        'postal_code',
      ],
    },
  ],
  status: 'OK',
});

const stores = fromJS([
  {
    storeId: '805',
    inventory: '0',
    stockStatus: 'Out of Stock',
    type: '',
    message: '',
    showPickupInStore: 'false',
    storeSelected: 'false',
  },
]);

const storeDetails = {
  allStores: [
    {
      type: 'Feature',
      geometry: {
        type: 'Point',
        coordinates: [
          -73.9783318,
          40.7576495,
        ],
      },
      properties: {
        Store_Zip: '10020',
        Sat_Open: '09:00',
        Country_Id: 'US',
        Sun_Close: '19:00',
        Mon_Close: '21:00',
        Tue_Open: '09:00',
        Shoes: '1',
        Thu_Open: '09:00',
        Store_City: 'New York',
        Store_Name: 'Between 48th & 49th',
        LoftPlus: 'N',
        Fri_Open: '09:00',
        Wed_Close: '21:00',
        Store_Phone: '(212) 922-3621',
        Store_Hours: 'Sunday 10:00 a.m. - 07:00 p.m.<br/>Monday 09:00 a.m. - 09:00 p.m.<br/>Tuesday 09:00 a.m. - 09:00 p.m.<br/>Wednesday 09:00 a.m. - 09:00 p.m.<br/>Thursday 09:00 a.m. - 09:00 p.m.<br/>Friday 09:00 a.m. - 09:00 p.m.<br/>Saturday 09:00 a.m. - 09:00 p.m.',
        Wed_Open: '09:00',
        Store_State: 'NY',
        distance: '462.5844240238359',
        Fri_Close: '21:00',
        Petites: '1',
        Store_Id: '805',
        Thu_Close: '21:00',
        Store_Address1: '600 5th Avenue',
        Company: 'ATS',
        Sun_Open: '10:00',
        Tue_Close: '21:00',
        Mon_Open: '09:00',
        Time_Zone: 'EST',
        Sat_Close: '21:00',
      },
    },
  ],
};

window.google = {
  maps: {
    InfoWindow: class {},
    LatLng: jest.fn(),
    MapTypeId: {},
    Map: (a, b) => ({ addListener: jest.fn() }),
    Marker: () => ({ addListener: jest.fn() }),
  },
};

describe('<GoogleMapVanilla />', () => {
  let GoogleMapComponent = '';
  beforeEach(() => {
    GoogleMapComponent = shallow(<GoogleMapVanilla
      google
      stores={stores}
      storeDetails={storeDetails}
      lat={location.getIn(['geometry', 'location', 'lat'])}
      lng={location.getIn(['geometry', 'location', 'lng'])}
    />);
  });

  test('should render correctly', () => {
    expect(GoogleMapComponent).toMatchSnapshot();
  });

  test('should not render when no allStores', () => {
    GoogleMapComponent = shallow(<GoogleMapVanilla
      google
      stores={stores}
      storeDetails={{}}
      lat={location.getIn(['geometry', 'location', 'lat'])}
      lng={location.getIn(['geometry', 'location', 'lng'])}
    />);
    expect(GoogleMapComponent).toMatchSnapshot();
  });

  test('should create ref when the component is mounted', () => {
    GoogleMapComponent = mount(<GoogleMapVanilla
      stores={stores}
      storeDetails={storeDetails}
      lat={location.getIn(['geometry', 'location', 'lat'])}
      lng={location.getIn(['geometry', 'location', 'lng'])}
      deviceType={DESKTOP}
      activeBrand={OUTLET_LOFT}
      isStoreLocator
    />);
    expect(GoogleMapComponent).toMatchSnapshot();
  });

  test('should close active info window when map clicked', () => {
    const infoWindow = { close: jest.fn() };
    GoogleMapComponent.setState({ activeInfoWindow: infoWindow });
    const onMapClicked = jest.spyOn(GoogleMapComponent.instance(), 'onMapClicked');
    onMapClicked();
    expect(GoogleMapComponent.state().activeInfoWindow).toEqual(null);
  });

  test('should call onMarkerClick', () => {
    const infoWindow = { open: jest.fn() };
    const onMarkerClick = jest.spyOn(GoogleMapComponent.instance(), 'onMarkerClick');
    onMarkerClick({
      isStoreLocator: true,
      store: storeDetails,
      map: {},
      marker: {},
      infoWindow,
    });
    expect(GoogleMapComponent.state().activeInfoWindow).toEqual(infoWindow);
  });
});
/* eslint-enable */
