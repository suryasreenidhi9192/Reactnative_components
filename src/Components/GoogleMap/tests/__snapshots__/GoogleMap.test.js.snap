// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GoogleMapVanilla /> should create ref when the component is mounted 1`] = `
<GoogleMap
  activeBrand="OUTLET_LOFT"
  deviceType="desktop"
  isStoreLocator={true}
  storeDetails={
    Object {
      "allStores": Array [
        Object {
          "geometry": Object {
            "coordinates": Array [
              -73.9783318,
              40.7576495,
            ],
            "type": "Point",
          },
          "properties": Object {
            "Company": "ATS",
            "Country_Id": "US",
            "Fri_Close": "21:00",
            "Fri_Open": "09:00",
            "LoftPlus": "N",
            "Mon_Close": "21:00",
            "Mon_Open": "09:00",
            "Petites": "1",
            "Sat_Close": "21:00",
            "Sat_Open": "09:00",
            "Shoes": "1",
            "Store_Address1": "600 5th Avenue",
            "Store_City": "New York",
            "Store_Hours": "Sunday 10:00 a.m. - 07:00 p.m.<br/>Monday 09:00 a.m. - 09:00 p.m.<br/>Tuesday 09:00 a.m. - 09:00 p.m.<br/>Wednesday 09:00 a.m. - 09:00 p.m.<br/>Thursday 09:00 a.m. - 09:00 p.m.<br/>Friday 09:00 a.m. - 09:00 p.m.<br/>Saturday 09:00 a.m. - 09:00 p.m.",
            "Store_Id": "805",
            "Store_Name": "Between 48th & 49th",
            "Store_Phone": "(212) 922-3621",
            "Store_State": "NY",
            "Store_Zip": "10020",
            "Sun_Close": "19:00",
            "Sun_Open": "10:00",
            "Thu_Close": "21:00",
            "Thu_Open": "09:00",
            "Time_Zone": "EST",
            "Tue_Close": "21:00",
            "Tue_Open": "09:00",
            "Wed_Close": "21:00",
            "Wed_Open": "09:00",
            "distance": "462.5844240238359",
          },
          "type": "Feature",
        },
      ],
    }
  }
  stores={
    Immutable.List [
      Immutable.Map {
        "storeId": "805",
        "inventory": "0",
        "stockStatus": "Out of Stock",
        "type": "",
        "message": "",
        "showPickupInStore": "false",
        "storeSelected": "false",
      },
    ]
  }
>
  <div
    style={
      Object {
        "minHeight": "100%",
        "minWidth": "100%",
      }
    }
  />
</GoogleMap>
`;

exports[`<GoogleMapVanilla /> should not render when no allStores 1`] = `""`;

exports[`<GoogleMapVanilla /> should render correctly 1`] = `
<div
  style={
    Object {
      "minHeight": "100%",
      "minWidth": "100%",
    }
  }
/>
`;
